addons/wallet/models/gift_card.py

```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8 -*-
+
+from uuid import uuid4
+from odoo import models, fields, api, registry, _
+from odoo.exceptions import ValidationError
+from odoo.osv import expression
+
+
+class GiftCard(models.Model):
+    # inherited Fields
+    _name = "gift.card"
+    _description = "Gift Card"
+    _order = 'id desc'
+    _check_company_auto = True
+    _sql_constraints = [
+        ('unique_gift_card_code', 'UNIQUE(code)',
+         'The gift card code must be unique.'),
+        ('check_amount', 'CHECK(amount >= 0)',
+         'The initial amount must be positive.')
+    ]
+
+    # helpers
+    @api.model
+    def _generate_code(self):
+        return str(uuid4())[4:-8]
+
+    # Relational fields
+    company_id = fields.Many2one(
+        'res.company', string='Company', index=True,
+        default=lambda self: self.env.company)
+
+    currency_id = fields.Many2one('res.currency',
+                                  string='Currency',
+                                  readonly=True,
+                                  related='company_id.currency_id')
+    sale_order_id = fields.Many2one("sale.order",
+                                    help="Sale Order where this gift card has been bought",
+                                    readonly=True)
+    wallet_transaction_id = fields.Many2one('wallet.transaction',
+                                            help="Wallet Transaction that add this gift card to a wallet")
+    partner_id = fields.Many2one('res.partner', help="If filled, only this customer can use it")
+
+    # simple fields
+    name = fields.Char(compute='_compute_name')
+    code = fields.Char(default=_generate_code, required=True, readonly=True, copy=False)
+    amount = fields.Monetary(required=True, currency_field='currency_id')
+    expired_date = fields.Date(copy=False,
+                               default=lambda self: fields.Date.add(fields.Date.context_today(self), years=1))
+    state = fields.Selection(
+        selection=[('new', 'New'), ('used', 'Used'), ('expired', 'Expired')],
```

Comment:
Yes The expiration date is here just to prevent the user from adding an expired gift card. (like coupon logic)

Reply:
Yes but it can be achieved without having a `computed` `state` field.
Just store `new` and `used`, and when applying the code just check the date to raise the error.
That will remove some code, the compute and the search.
Worst case we have coupon in `new` state with an expired date that is expired, but this can be fixed if you want with a cron as in coupon `cron_expire_coupon`.
It makes the `state` field way easier to read/handle -> no more compute, no more _search.
Set to `new` on create, to `used` when setting `wallet_transaction_id`, and `expired` with a cron (if you want)

Reply:
Yes I think this way is more optimized (because we have to call the cron once a day (instead of calling it every time)